### DISCUSSION TITLE
Usability of UnusedPsalmSuppress + DeprecatedMethod where multi-version compatibility is desired

### DISCUSSION THREAD

**Discussion by weirdan** [vimeo/psalm#8968] [Q&A]
Suppose we have the following code:

```php
<?php

class C {
    // /** @deprecated */ 
    // public function doThings(): void {} // deprecated in 1.5 and removed in 2.0
}

// this code needs to work with 1.x and 2.x
// but it emits UnusedPsalmSuppress
$o = new C;
if (method_exists($o, 'doThings')) {
    // for, e.g., v1.1, the method is not deprecated - UnusedPsalmSuppress is reported
    // for 1.5+ it is
    // for 2.0+, it is absent but added as a phantom method
    // phantom methods have no docblocks, and thus the method is not deprecated again
    // so UnusedPsalmSuppress is reported again
    /** @psalm-suppress DeprecatedMethod */
    $o->doThings();
}
```
(https://psalm.dev/r/58970303df)

where `C` comes from an external library and `C::doThings()` method was deprecated in v1.5 of the aforementioned library and subsequently removed in v2.0. Also, assume that Psalm is configured to report unused suppressions.

Given the above, do we have any other option apart from suppressing/baselining UnusedPsalmSuppress?

Real-world example: https://github.com/psalm/psalm-plugin-symfony/pull/296/files#diff-c346b1aeac8259782c74702eadfe1d7a240ee3ecea6020ad2addede93c635cab

**Comment by orklah** [comment]
suppressing through config seems best, or otherwise disabling the find unused suppress I guess...

**Reply by AndrolGenhald** [reply]
Since it's inside a `method_exists` check for a deprecation, I'd personally just do the suppression inline. Once the code requires 1.5+ that block will be removed anyway, and that way there's no chance of leaving behind unneeded configs. Unless you mean to just blanket suppress it for the entire codebase via config, that's also a viable option if there are a bunch of places this happens and it's more annoyance than it's worth.

**Reply by weirdan** [reply]
I tried inline suppression (`/** @psalm-suppress DeprecatedMethod, UnusedPsalmSuppress **/`) but that didn't work (in Psalm 4.30), so I resorted to baseline suppression instead.
